addon/helpers/-element.js

```diff
@@ -20,15 +20,8 @@ export default Helper.extend({
       this.tagName = tagName;
 
       if (typeof tagName === 'string') {
-        // return a different component name to force a teardown
-        if (this.componentName === '-dynamic-element') {
-          this.componentName = '-dynamic-element-alt';
-        } else {
-          this.componentName = '-dynamic-element';
-        }
+        return Component.extend();
```

Comment:
This is going to give you a different component instance every time this helper recomputes. It's not stable. *Everything* below this point in the render hierarchy will get destroyed and re-created.

Instead, we really only want it to change when the `tagName` argument changes.

Reply:
Oh, I see that higher in the code this is guarded by `tagName !== this.tagName`, so I think my above comment is wrong.

Reply:
> so I think my above comment is wrong.

I hope so! :)

However your comment made me aware and worry a bit that we have the `.extend()` in the hot path here. Reusing the same component class every time does not work though when the tag name changes, as the component instance is not invalidated, I guess for the same reason there were two separate -dynamic-element components before. But maybe we should also return two alternating component classes, also because the `ensure-safe-component` helpers will register a new component each and every time IIUC: https://github.com/embroider-build/embroider/blob/master/packages/util/addon/index.js#L45-L53

Not sure why anyone would change tag names for the same instance, but I guess we can still cover this case better?

Reply:
Yes, I think it would be good to keep two alternating classes. The ensureSafeComponent function definitely makes that easier now, since they can easily be defined inline in this file, instead of being globally exposed components.

Reply:
In the second commit I added back the alternating component class pattern, however as you said defined locally without exposing them.

A bunch of tests fail though, and the interesting part here is that *all* of them pass when running in isolation. I think the reason is that [this WeakMap)(https://github.com/embroider-build/embroider/blob/master/packages/util/addon/index.js#L42) survives across tests, although the owner changes. So any following test would reuse the map, assume that the component from the previous test run is already registered with the owner, but which is not the case as the owner changed.

I prepared a PR for embroider, that fixes this problem locally (when using it with `yarn link`): https://github.com/embroider-build/embroider/pull/624